### PR TITLE
Allow apiserver to listen on 0.0.0.0

### PR DESCRIPTION
**Pull request by kazegusuri**
IP addresses of ReadOnly Port  and Secure Port  can be specified by `--public_address_override` option. I want to set 0.0.0.0 to access apiserver from both public address and localhost. If 0.0.0.0 is specified via `--public_address_override`,  endpoints of `kubernetes` and `kubernetes_ro` services are also set 0.0.0.0 like following:

```
$ kubectl get endpoints 
NAME                ENDPOINTS
kubernetes          0.0.0.0:6443
kubernetes-ro       0.0.0.0:7080
```

This PR adds `--listen_all_interfaces` option to listen on 0.0.0.0 or not. Services' endpoints still refer address specified by --public_address_override`.
There are some alternatives to fix the problem, so comments are welcome.

### CODE REVIEW

**Comment by roberthbailey** [comment]
This annoys me all the time as well. Is there a reason not to always bind the SSL port to all interfaces? (I can see an argument not to bind unsecured ports).

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -125,6 +127,8 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 		"Read only port will be opened on this address, and it is assumed that port "+
 		"443 at this address will be proxied/redirected to '-address':'-port'. If "+
 		"blank, the address in the first listed interface will be used.")
+	fs.BoolVar(&s.ListenAllInterfaces, "listen_all_interfaces", s.ListenAllInterfaces, ""+
```
Comment (thockin) [inline cmd/kube-apiserver/app/server.go]: Naively this overlaps completely with --address - why doesn't that work?

Perhaps we should rename flags and fields until we can look at this logic and know what it does?

Comment (kazegusuri) [inline cmd/kube-apiserver/app/server.go]: --address affects only Localhost port, but --public_address_override affects ReadOnly/Secure ports and also endpoints of kubernetes services.
I want to set address for ReadOnly/Secure ports independently from endpoints.

**Comment by kazegusuri** [comment]
Always binding all interfaces is acceptable for me.

**Comment by roberthbailey** [comment]
We certainly don't want to bind the http read/write port (currently 8080) to all interfaces by default.

**Comment by thockin** [comment]
I am re-discovering this logic and it's a mess.

We have several different ideas and we conflate them together at every
opportunity.

We have Address (--address) which defaults to 127.0.0.1.  This hosts the
WideOpenPort (--port) which has neither SSL not authn/authz.

We have PublicAddress which defaults to ChooseHostInterface() but can be
overridden (--public_address_override).  This hosts the ReadOnlyPort
(--read_only_port) and SecurePort (--secure_port).  It is also what we use
to generate a self-signed cert for client authn.

We have clusterIP which is set from PublicAddress and is used to publish
the master service.


I think some relatively easy cleanup will make this more obvious.  PR #6526
maybe?

**Comment by thockin** [comment]
Replaced by #6526  I think.

**Pull request closed without merging**